### PR TITLE
CZIs without Scene 

### LINKED ISSUE #67: Coordinates Overspecified Exception with certain Mosaic files
I discovered it when I tried opening a tilescan with _aicsimageio_. So, I created a library with any type of possible combinations of dimensions: M, S, T and Z with two channels.

Versions of packages I was using:

python: 3.8.5
aicspylibczi: 2.7.6
aicsimageio: 3.3.3
czifile: 2019.7.2

If I try to open tilescans (all of them with two channels) with only tiles, tiles and time, tiles and z, or tiles, time and z I get an error in line 57 of the file aicspylibczi/CziFile.py: __aicspylibczi.PylibCZI_CDimCoordinatesOverspecifiedException_

However, if I have positions (S) as an additional dimension I don't get this error, but I do get the same behaviour as already specified in the issue here: [AllenCellModeling/aicsimageio/issues/163].

I tried using CziFile in terminal, but also from scripts, starting with AICSImage or CziReader. In contrary, if I use the package czifile everything works smoothly and I can load any of the files.

I can provide the file library I created via E-Mail.

_Originally posted by @dimi-huer in https://github.com/AllenCellModeling/aicspylibczi/issues/62#issuecomment-745613504_

### PR DESCRIPTION
**Pull request by heeler**
This PR resolves #67. I took an initial quick shot at fixing it and then the issue got conflated with another problem and it took me a while to realize they were independent issues. The PR addresses the problem of CZI files without a scene. 

Details:
1) The prior code was dependent on the file having a scene and this was a logical error. This has now been corrected. 
2) The Reader.* files have also been cleaned up a bit (clang-format and clang-tidy). 

The test files were provided by @dimi-huer and were invaluable in resolving the problem and I am very grateful to him for bringing this issue to our attention.

I've added comments where the key changes are. 
Additionally, I ran clang-format to clean up the spacing I suggest viewing the changed files with the appended `?w=1` which [suppresses](https://github.com/AllenCellModeling/aicspylibczi/pull/69/files?w=1) much of the white space highlighting.

### CODE REVIEW

File: _aicspylibczi/Reader.cpp [hunk 1]
```diff
@@ -1,469 +1,481 @@
...
+  if ((!sceneBool || sceneSize == 1)) {
+    // scenes are not defined so the dimBounds define the shape
+    m_statistics.dimBounds.EnumValidDimensions([&tbl](libCZI::DimensionIndex di_, int start_, int size_) -> bool {
+      tbl.emplace(dimensionIndexToDimIndex(di_),
+                  std::make_pair(start_, size_ + start_)); // changed from [start, end) to be [start, end]
+      return true;
+    });
+    if (isMosaic()) {
+      tbl.emplace(charToDimIndex('M'), std::make_pair(m_statistics.minMindex, m_statistics.maxMindex + 1));
     }
```
Comment (heeler) [inline _aicspylibczi/Reader.cpp:161]: Enable the no scene branch to handle mosaic files

File: _aicspylibczi/Reader.cpp [hunk 2]
```diff
@@ -1,469 +1,481 @@
...
+Reader::DimIndexRangeMap
+Reader::sceneShape(int scene_index_)
+{
+  bool sceneBool(false);
+  int sceneStart(0), sceneSize(0);
+  tie(sceneBool, sceneStart, sceneSize) = scenesStartSize();
+
+  DimIndexRangeMap tbl;
+
+  if ((!sceneBool || sceneSize == 1)) {
```
Comment (heeler) [inline _aicspylibczi/Reader.cpp:152]: Allow this code to handle mosaic files.

**Review by JacksonMaxfield — APPROVED** [review]
I think this all looks fine? I still need to learn C++ to really understand what is an isn't good style / practice.

Because of that, I am approving basically because I see the test and that makes sense to me and the tests pass so seems good. But I would wait for @toloudis approval.

File: _aicspylibczi/Reader.cpp [hunk 3]
```diff
@@ -1,469 +1,481 @@
...
+    if (isMosaic()) {
+      tbl.emplace(charToDimIndex('M'), std::make_pair(m_statistics.minMindex, m_statistics.maxMindex + 1));
     }
+    auto xySize = getSceneYXSize();
+    tbl.emplace(charToDimIndex('Y'), std::make_pair(0, xySize.h));
+    tbl.emplace(charToDimIndex('X'), std::make_pair(0, xySize.w));
+  } else {
+    if (scene_index_ < sceneStart || sceneStart + sceneSize <= scene_index_) {
+      std::stringstream ss;
+      ss << "Scene index " << scene_index_ << " ∉ "
```
Comment (toloudis) [inline _aicspylibczi/Reader.cpp:168]: what's that funky unicode symbol?

Comment (toloudis) [reply _aicspylibczi/Reader.cpp:168]: is that "not an element of" ?

File: _aicspylibczi/Reader.h [hunk 4]
```diff
@@ -66,265 +66,262 @@
...
+  bool shapeIsConsistent() const { return !m_specifyScene; }
+
+  virtual ~Reader() { m_czireader->Close(); }
+
+  /*!
+   * Get the full size of the mosaic image without scaling. If you're selecting a sub-region it must be within the box
+   * returned.
+   * @return an IntRect {x, y, w, h}
+   */
+  libCZI::IntRect mosaicShape() const { return m_statistics.boundingBoxLayer0Only; }
```
Comment (toloudis) [inline _aicspylibczi/Reader.h:270]: random thing i am noticing: is this guaranteed accurate?  I sort of recall it as being a short cut (i mean using boundingBoxLayer0Only)

File: _aicspylibczi/Reader.h [hunk 5]
```diff
@@ -66,265 +66,262 @@
...
+
+private:
+  Reader::SubblockIndexVec getMatches(SubblockSortable& match_);
+
+  static bool isPyramid0(const libCZI::SubBlockInfo& info_)
+  {
+    return (info_.logicalRect.w == info_.physicalSize.w && info_.logicalRect.h == info_.physicalSize.h);
+  }
+
+  // void getMemory(SubblockIndexVec& matches_);
```
Comment (toloudis) [inline _aicspylibczi/Reader.h:317]: can remove?

File: c_tests/test_Reader.cpp [hunk 6]
```diff
@@ -476,10 +552,10 @@ TEST_CASE_METHOD(CziBgrCreator2, "test_bgr2_flatten", "[Reader_read_flatten_bgr2
 
   REQUIRE(pr.size() == 1);
 
-  for (auto x : pr) {
+  for (const auto& x : pr) {
     REQUIRE(x->shape()[0] == 3);
     REQUIRE(x->shape()[1] == 81);
-    REQUIRE(x->shape()[2] == 147);
+    REQUIRE(147 == x->shape()[2]);
```
Comment (toloudis) [inline c_tests/test_Reader.cpp:558]: this line got flipped but the others didn't?

**Review by toloudis — APPROVED** [review]
LGTM!  Despite muddying the PR diffs, I am really glad the formatting was done on this code.

File: _aicspylibczi/Reader.cpp (hunk 3 above)
Comment (heeler) [reply _aicspylibczi/Reader.cpp:168]: exactly (set notation)

File: _aicspylibczi/Reader.h (hunk 4 above)
Comment (heeler) [reply _aicspylibczi/Reader.h:270]: I'm not sure what you mean here. `boundingBoxLayer0Only` just means don't give me the bounding box of Non-Zero pyramid levels ie give me the data that was acquired not a pyramid level that was computationally generated.
To the best of my knowledge, there is no reason to believe `boundingBoxLayer0Only` would be inaccurate

File: _aicspylibczi/Reader.h (hunk 5 above)
Comment (heeler) [reply _aicspylibczi/Reader.h:317]: Thanks I don't know how I always miss these.

File: _aicspylibczi/Reader.cpp (hunk 3 above)
Comment (JacksonMaxfield) [reply _aicspylibczi/Reader.cpp:168]: Can you comment that.

File: _aicspylibczi/Reader.h (hunk 4 above)
Comment (heeler) [reply _aicspylibczi/Reader.h:270]: I reread the docs and though they are thin I believe the bounding box will give the dims of the mosaic file. 
The previous issue was using boundBoxLayer0Only or other libCZI::IntRects in calculations resulted in a loss of precision and overflow in the calculation

**Pull request merged**